stablehlo/transforms/StablehloLegalizeDeprecatedOps.cpp

```diff
@@ -0,0 +1,162 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// Implements composite inlining.
+
+#include <cassert>
+#include <cstdint>
+
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/PatternMatch.h"
+#include "stablehlo/dialect/StablehloOps.h"
+#include "stablehlo/transforms/Passes.h"
+
+namespace mlir {
+namespace stablehlo {
+
+#define GEN_PASS_DEF_STABLEHLOLEGALIZEDEPRECATEDOPSPASS
+#include "stablehlo/transforms/Passes.h.inc"
+
+namespace {
+
+struct StablehloLegalizeDeprecatedOpsPass final
+    : impl::StablehloLegalizeDeprecatedOpsPassBase<
+          StablehloLegalizeDeprecatedOpsPass> {
+  StablehloLegalizeDeprecatedOpsPass()
+      : StablehloLegalizeDeprecatedOpsPassBase<
+            StablehloLegalizeDeprecatedOpsPass>() {};
+  StablehloLegalizeDeprecatedOpsPass(
+      const StablehloLegalizeDeprecatedOpsPassOptions &opts)
+      : StablehloLegalizeDeprecatedOpsPassBase<
+            StablehloLegalizeDeprecatedOpsPass>(opts) {};
+
+  LogicalResult initialize(MLIRContext *context) override {
+    target = std::make_shared<ConversionTarget>(*context);
+    target->addIllegalOp<BroadcastOp, CreateTokenOp, CrossReplicaSumOp, DotOp,
+                         UnaryEinsumOp>();
+
+    if (failOnUnusedOps) {
+      // Deprecated ops to be removed with no replacements
+      target->addIllegalOp<MapOp, RngOp, TraceOp>();
+    }
+
+    target->addLegalDialect<StablehloDialect>();
+
+    RewritePatternSet patterns_(context);
+    populateStablehloLegalizeDeprecatedOpsPatterns(context, &patterns_);
+    patterns = std::move(patterns_);
+
+    return success();
+  }
+
+  void runOnOperation() override {
+    if (failed(applyPartialConversion(getOperation(), *target, patterns))) {
+      return signalPassFailure();
+    }
+  }
+
+ private:
+  std::shared_ptr<ConversionTarget> target;
+  FrozenRewritePatternSet patterns;
+};
+
+///////////
+// Patterns
+///////////
+
+struct CrossReplicaSumToAllReducePattern
+    : public OpRewritePattern<CrossReplicaSumOp> {
+  using OpRewritePattern<CrossReplicaSumOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(CrossReplicaSumOp crossReplicaSumOp,
+                                PatternRewriter &rewriter) const override {
+    auto allReduceOp = rewriter.replaceOpWithNewOp<AllReduceOp>(
+        crossReplicaSumOp, crossReplicaSumOp.getType(),
+        crossReplicaSumOp.getOperand(), crossReplicaSumOp.getReplicaGroups(),
+        /*channel_handle=*/ChannelHandleAttr(),
```

Comment:
I'm a bit confused by this. The spec for `all_reduce` says:

```
The operation splits the StableHLO process grid into process_groups which is defined as follows:
cross_replica(replica_groups) if channel_id <= 0 and use_global_device_ids = false.
```

Does `ChannelHandleAttr` somehow represent `channel_id <= 0`?

Reply:
Migrated from: https://github.com/openxla/xla/blob/bd83c512acb8b68a82ba21b3180253d958f20513/xla/mlir_hlo/mhlo/transforms/legalize_cross_replica_sum_to_all_reduce/legalize_cross_replica_sum_to_all_reduce.cc#L41

Channel handle is an optional attr, and an null attribute builds this with the optional as unset. Tried passing `nullopt` but that's causes a compilation error.